### PR TITLE
Tozpatron 300 volunteer applications

### PR DESCRIPTION
**Pull request by macieg-b**
CRUD for proposals

### CODE REVIEW

File: src/main/java/com/intive/patronage/toz/proposals/ProposalController.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    })
+    @DeleteMapping(value = "/{id}")
+    @PreAuthorize("hasAnyAuthority('SA', 'TOZ')")
+    public ResponseEntity<?> deleteProposal(@PathVariable UUID id) {
+        proposalService.delete(id);
+        return ResponseEntity.ok().build();
+    }
+
+    private void validateProposalRoles(ProposalView proposalView){
+        for(Role role : proposalView.getRoles()){
```
Comment (maras93) [inline src/main/java/com/intive/patronage/toz/proposals/ProposalController.java]: Please use auto reformat code.

File: src/test/java/com/intive/patronage/toz/mail/MailTemplatesServiceTest.java [hunk 2]
```diff
@@ -1,12 +1,20 @@
...
 import org.junit.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.TestPropertySource;
 
 import java.io.IOException;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+@SpringBootTest
+@TestPropertySource(
```
Comment (maras93) [inline src/test/java/com/intive/patronage/toz/mail/MailTemplatesServiceTest.java]: I think this is unnecessary for these tests.

File: src/main/java/com/intive/patronage/toz/proposals/ProposalController.java [hunk 3]
```diff
@@ -1,4 +1,101 @@
...
+    @ApiOperation(value = "Delete proposal", notes =
+            "Required roles: SA, TOZ.")
+    @ApiResponses(value = {
+            @ApiResponse(code = 404, message = "Proposal not found", response = ErrorResponse.class)
+    })
+    @DeleteMapping(value = "/{id}")
+    @PreAuthorize("hasAnyAuthority('SA', 'TOZ')")
+    public ResponseEntity<?> deleteProposal(@PathVariable UUID id) {
+        proposalService.delete(id);
+        return ResponseEntity.ok().build();
```
Comment (Jakub-Izbicki) [inline src/main/java/com/intive/patronage/toz/proposals/ProposalController.java]: It should return deleted object.

File: src/main/java/com/intive/patronage/toz/proposals/ProposalService.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    Proposal create(final Proposal proposal) {
+        final String email = proposal.getEmail();
+        if (proposalRepository.existsByEmail(email)) {
+            throw new AlreadyExistsException(PROPOSAL);
+        }
+        if (userRepository.existsByEmail(email)) {
+            throw new AlreadyExistsException(USER);
+        }
```
Comment (Jakub-Izbicki) [inline src/main/java/com/intive/patronage/toz/proposals/ProposalService.java]: These two validations could be extracted to another method, e.g. `validateProposalEmail()`

File: src/test/java/com/intive/patronage/toz/proposals/ProposalServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    @Test
+    @UseDataProvider("getProperProposal")
+    public void shouldCreateProposal(final Proposal proposal) throws Exception {
+        when(proposalRepository.save(any(Proposal.class))).thenReturn(proposal);
+
+        final Proposal createdProposal = proposalService.create(proposal);
+        assertEquals(EXPECTED_NAME, createdProposal.getName());
+        assertEquals(EXPECTED_MAIL, createdProposal.getEmail());
+        assertEquals(EXPECTED_SURNAME, createdProposal.getSurname());
+        assertTrue(createdProposal.hasRole(EXPECTED_ROLE));
```
Comment (Jakub-Izbicki) [inline src/test/java/com/intive/patronage/toz/proposals/ProposalServiceTest.java]: AssertJ could be used here: `assertThat(createdProposal.getRoles()).contains(EXPECTED_ROLE);`

File: src/main/java/com/intive/patronage/toz/users/model/db/Role.java [hunk 6]
```diff
@@ -3,7 +3,7 @@
 import org.springframework.security.core.GrantedAuthority;
 
 public enum Role implements GrantedAuthority {
-    SA, TOZ, VOLUNTEER, ANONYMOUS;
+    SA, TOZ, VOLUNTEER, TEMP_HOUSE, ANONYMOUS;
```
Comment (Jakub-Izbicki) [inline src/main/java/com/intive/patronage/toz/users/model/db/Role.java]: I am not sure if `TEMP_HOUSE` as an auth role is a good decision. First, How does it really differ from `VOLUNTEER` role? Second, I thought that this purpose of volunteer (whether he is a normal volunteer, or a temporary house) has nothing to do with authentication, and it had only an informational value to the admin. We can discuss it at a meeting if you want.

File: src/test/java/com/intive/patronage/toz/mail/MailTemplatesServiceTest.java (hunk 2 above)
Comment (macieg-b) [inline src/test/java/com/intive/patronage/toz/mail/MailTemplatesServiceTest.java]: I don't know why but without this environment value it doesn't work so I had to add it.

File: src/main/java/com/intive/patronage/toz/proposals/ProposalController.java (hunk 3 above)
Comment (macieg-b) [inline src/main/java/com/intive/patronage/toz/proposals/ProposalController.java]: It's not a rule. We don't return deleted object in for example Pet and User controllers.

File: src/main/java/com/intive/patronage/toz/proposals/ProposalService.java (hunk 4 above)
Comment (macieg-b) [inline src/main/java/com/intive/patronage/toz/proposals/ProposalService.java]: Done

File: src/test/java/com/intive/patronage/toz/proposals/ProposalServiceTest.java (hunk 5 above)
Comment (macieg-b) [inline src/test/java/com/intive/patronage/toz/proposals/ProposalServiceTest.java]: Done

File: src/main/java/com/intive/patronage/toz/users/model/db/Role.java (hunk 6 above)
Comment (macieg-b) [inline src/main/java/com/intive/patronage/toz/users/model/db/Role.java]: We don't know if TEMP_HOUSE will have different privileges in future.

File: src/test/java/com/intive/patronage/toz/mail/MailTemplatesServiceTest.java (hunk 2 above)
Comment (maras93) [inline src/test/java/com/intive/patronage/toz/mail/MailTemplatesServiceTest.java]: You should remove either `@TestPropertySource` and `@SpringBootTest`.

File: src/main/java/com/intive/patronage/toz/users/model/db/Role.java (hunk 6 above)
Comment (mlotysz) [inline src/main/java/com/intive/patronage/toz/users/model/db/Role.java]: `TEMP_HOUSE` should be removed. Temporary house will be done in another PR/story.

Comment (macieg-b) [inline src/main/java/com/intive/patronage/toz/users/model/db/Role.java]: @mlotysz https://tracker.intive.com/jira/browse/TOZPATRON-158

File: src/test/java/com/intive/patronage/toz/mail/MailTemplatesServiceTest.java (hunk 2 above)
Comment (macieg-b) [inline src/test/java/com/intive/patronage/toz/mail/MailTemplatesServiceTest.java]: You are right, it works by now.
